File: spring-data-requery/src/main/java/org/springframework/data/requery/core/RequeryOperations.java
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.core;

import io.requery.TransactionIsolation;
import io.requery.meta.Attribute;
import io.requery.meta.EntityModel;
import io.requery.meta.QueryAttribute;
import io.requery.query.Condition;
import io.requery.query.Deletion;
import io.requery.query.Expression;
import io.requery.query.InsertInto;
import io.requery.query.Insertion;
import io.requery.query.Result;
import io.requery.query.Scalar;
import io.requery.query.Selection;
import io.requery.query.Tuple;
import io.requery.query.Update;
import io.requery.query.element.QueryElement;
import io.requery.query.function.Count;
import io.requery.sql.EntityContext;
import io.requery.sql.EntityDataStore;
import org.springframework.data.requery.mapping.RequeryMappingContext;
import org.springframework.data.requery.utils.Iterables;
import org.springframework.data.requery.utils.RequeryUtils;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.function.Function;

import static org.springframework.data.requery.utils.RequeryUtils.unwrap;

/**
 * Java용 RequeryOperations
 *
 * @author debop
 * @since 18. 6. 4
 */
@ParametersAreNonnullByDefault
public interface RequeryOperations {

    @Nonnull
    EntityDataStore<Object> getDataStore();

    @Nonnull
    RequeryMappingContext getMappingContext();

    default EntityModel getEntityModel() {
        return RequeryUtils.getEntityModel(getDataStore());
    }

    @SuppressWarnings("unchecked")
    default <E> EntityContext<E> getEntityContext() {
        return RequeryUtils.getEntityContext(getDataStore());
    }

    default <E> Selection<? extends Result<E>> select(@Nonnull final Class<E> entityType) {
        return getDataStore().select(entityType);
    }

    default <E> Selection<? extends Result<E>> select(@Nonnull final Class<E> entityType,
                                                      @Nonnull final QueryAttribute<?, ?>... attributes) {
        return getDataStore().select(entityType, attributes);
    }

    default Selection<? extends Result<Tuple>> select(@Nonnull final Expression<?>... expressions) {
        return getDataStore().select(expressions);
    }

    default <E, K> E findById(@Nonnull final Class<E> entityType, @Nonnull final K id) {
        return getDataStore().findByKey(entityType, id);
    }

    default <E> List<E> findAll(@Nonnull final Class<E> entityType) {
        return getDataStore().select(entityType).get().toList();
    }

    default <E> E refresh(@Nonnull final E entity) {
        return getDataStore().refresh(entity);
    }

    @SuppressWarnings("UnusedReturnValue")
    default <E> E refresh(@Nonnull final E entity, @Nullable final Attribute<?, ?>... attributes) {
        return getDataStore().refresh(entity, attributes);
    }

    default <E> List<E> refresh(@Nonnull final Iterable<E> entities, final Attribute<?, ?>... attributes) {
        return Iterables.toList(getDataStore().refresh(entities, attributes));
    }

    default <E> E refreshAllProperties(@Nonnull final E entity) {
        return getDataStore().refreshAll(entity);
    }

    default <E> E upsert(@Nonnull final E entity) {
        return getDataStore().upsert(entity);
    }

    default <E> List<E> upsertAll(@Nonnull final Iterable<E> entities) {
        return Iterables.toList(getDataStore().upsert(entities));
    }

    default <E> E insert(@Nonnull final E entity) {
        return getDataStore().insert(entity);
    }

    default <E, K> K insert(@Nonnull final E entity, @Nonnull final Class<K> keyClass) {
        return getDataStore().insert(entity, keyClass);
    }

    default <E> Insertion<? extends Result<Tuple>> insert(@Nonnull final Class<E> entityType) {
        return getDataStore().insert(entityType);
    }

    @SuppressWarnings("unchecked")
    default <E> InsertInto<? extends Result<Tuple>> insert(@Nonnull final Class<E> entityType, QueryAttribute<E, ?>... attributes) {
        return getDataStore().insert(entityType, attributes);
    }

    default <E> List<E> insertAll(@Nonnull final Iterable<E> entities) {
        return Iterables.toList(getDataStore().insert(entities));
    }

    default <E, K> List<K> insertAll(@Nonnull final Iterable<E> entities, @Nonnull final Class<K> keyClass) {
        return Iterables.toList(getDataStore().insert(entities, keyClass));
    }

    @Nonnull
    default Update<? extends Scalar<Integer>> update() {
        return getDataStore().update();
    }

    default <E> E update(@Nonnull final E entity) {
        return getDataStore().update(entity);
    }

    default <E> E update(@Nonnull final E entity, final Attribute<?, ?>... attributes) {
        return getDataStore().update(entity, attributes);
    }

    default <E> Update<? extends Scalar<Integer>> update(@Nonnull final Class<E> entityType) {
        return getDataStore().update(entityType);
    }

    default <E> List<E> updateAll(@Nonnull final Iterable<E> entities) {
        return Iterables.toList(getDataStore().update(entities));
    }

    default Deletion<? extends Scalar<Integer>> delete() {
        return getDataStore().delete();
    }

    default <E> Deletion<? extends Scalar<Integer>> delete(@Nonnull final Class<E> entityType) {
        return getDataStore().delete(entityType);
    }

    default <E> void delete(@Nonnull final E entity) {
        getDataStore().delete(entity);
    }

    default <E> void deleteAll(@Nonnull final Iterable<E> entities) {
        getDataStore().delete(entities);
    }

    default <E> Integer deleteAll(@Nonnull final Class<E> entityType) {
        return getDataStore().delete(entityType).get().value();
    }

    default <E> Selection<? extends Scalar<Integer>> count(@Nonnull final Class<E> entityType) {
        return getDataStore().count(entityType);
    }

    default Selection<? extends Scalar<Integer>> count(final QueryAttribute<?, ?>... attributes) {
        return getDataStore().count(attributes);
    }

    @SuppressWarnings("unchecked")
    default <E> int count(@Nonnull final Class<E> entityType,
                          @Nonnull final QueryElement<? extends Result<E>> whereClause) {
        QueryElement<?> query = RequeryUtils.applyWhereClause(unwrap(select(Count.count(entityType))), whereClause.getWhereElements());
        Tuple tuple = ((QueryElement<? extends Result<Tuple>>) query).get().first();
        return tuple.<Integer>get(0);
    }

    default <E> boolean exists(@Nonnull final Class<E> entityType,
                               @Nonnull final QueryElement<? extends Result<E>> whereClause) {
        return whereClause.limit(1).get().firstOrNull() != null;
    }

    default <E, K> boolean existsBy(@Nonnull final Class<E> entityType,
                                    @Nonnull final Condition<K, ?> condition) {
        return getDataStore().count(entityType).from(entityType).where(condition).get().value() > 0;
    }

    default Result<Tuple> raw(@Nonnull final String query, final Object... parameters) {
        return getDataStore().raw(query, parameters);
    }

    default <E> Result<E> raw(@Nonnull final Class<E> entityType,
                              @Nonnull final String query,
                              final Object... parameters) {
        return getDataStore().raw(entityType, query, parameters);
    }

    default <V> V runInTransaction(@Nonnull final Callable<V> callable) {
        return runInTransaction(callable, null);
    }

    <V> V runInTransaction(@Nonnull final Callable<V> callable, @Nullable final TransactionIsolation isolation);

    default <V> V withTransaction(@Nonnull final Function<EntityDataStore<Object>, V> block) {
        return withTransaction(block, null);
    }

    <V> V withTransaction(@Nonnull final Function<EntityDataStore<Object>, V> block, @Nullable final TransactionIsolation isolation);
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/core/RequeryOperations.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.core

import io.requery.TransactionIsolation
import io.requery.kotlin.Deletion
import io.requery.kotlin.InsertInto
import io.requery.kotlin.Insertion
import io.requery.kotlin.QueryableAttribute
import io.requery.kotlin.Selection
import io.requery.kotlin.Update
import io.requery.meta.Attribute
import io.requery.meta.EntityModel
import io.requery.query.Expression
import io.requery.query.Result
import io.requery.query.Scalar
import io.requery.query.Tuple
import io.requery.query.element.QueryElement
import io.requery.query.function.Count
import io.requery.sql.EntityContext
import io.requery.sql.KotlinEntityDataStore
import org.springframework.data.requery.kotlin.applyWhereConditions
import org.springframework.data.requery.kotlin.getEntityContext
import org.springframework.data.requery.kotlin.getEntityModel
import org.springframework.data.requery.kotlin.mapping.RequeryMappingContext
import kotlin.reflect.KClass

/**
 *  Java용 RequeryOperations
 *  // TODO: Move to Java code
 *
 * @author debop
 */
interface RequeryOperations {

    val dataStore: KotlinEntityDataStore<Any>
    val mappingContext: RequeryMappingContext

    @JvmDefault
    val entityModel: EntityModel
        get() = dataStore.getEntityModel()

    @JvmDefault
    val entityContext: EntityContext<Any>
        get() = dataStore.getEntityContext()

    @JvmDefault
    fun <E : Any> select(entityType: KClass<E>): Selection<out Result<E>> = dataStore.select(entityType)

    @JvmDefault
    fun <E : Any> select(entityType: KClass<E>, vararg attributes: QueryableAttribute<E, *>): Selection<out Result<E>> =
        dataStore.select(entityType, *attributes)

    @JvmDefault
    fun select(vararg expressions: Expression<*>): Selection<out Result<Tuple>> = dataStore.select(*expressions)

    @JvmDefault
    fun <E : Any, K> findById(entityType: KClass<E>, id: K): E? = dataStore.findByKey(entityType, id)

    @JvmDefault
    fun <E : Any> findAll(entityType: KClass<E>): List<E> = dataStore.select(entityType).get().toList()

    @JvmDefault
    fun <E : Any> refresh(entity: E): E = dataStore.refresh(entity)

    @JvmDefault
    fun <E : Any> refresh(entity: E, vararg attributes: Attribute<*, *>): E =
        dataStore.refresh(entity, *attributes)

    @JvmDefault
    fun <E : Any> refreshAll(entities: Iterable<E>, vararg attributes: Attribute<*, *>): List<E> =
        dataStore.refresh<E>(entities, *attributes).toList()

    @JvmDefault
    fun <E : Any> refreshAllProperties(entity: E): E = dataStore.refreshAll(entity)

    @JvmDefault
    fun <E : Any> refreshAllEntities(entities: Iterable<E>, vararg attributes: Attribute<E, *>): List<E> =
        entities.map { refreshAllProperties(it) }

    @JvmDefault
    fun <E : Any> upsert(entity: E): E = dataStore.upsert(entity)

    @JvmDefault
    fun <E : Any> upsertAll(entities: Iterable<E>): List<E> =
        dataStore.upsert<E>(entities).toList()

    @JvmDefault
    fun <E : Any> insert(entity: E): E = dataStore.insert(entity)

    @JvmDefault
    fun <E : Any, K : Any> insert(entity: E, keyClass: KClass<K>): K = dataStore.insert(entity, keyClass)

    @JvmDefault
    fun <E : Any> insert(entityType: KClass<E>): Insertion<out Result<Tuple>> = dataStore.insert<E>(entityType)

    @JvmDefault
    fun <E : Any> insert(entityType: KClass<E>, vararg attributes: QueryableAttribute<E, *>): InsertInto<out Result<Tuple>> =
        dataStore.insert(entityType, *attributes)

    @JvmDefault
    fun <E : Any> insertAll(entities: Iterable<E>): List<E> = dataStore.insert<E>(entities).toList()

    @JvmDefault
    fun <E : Any, K : Any> insertAll(entities: Iterable<E>, keyClass: KClass<K>): List<K> =
        dataStore.insert<K, E>(entities, keyClass).toList()

    @JvmDefault
    fun update(): Update<out Scalar<Int>> = dataStore.update()

    @JvmDefault
    fun <E : Any> update(entity: E): E = dataStore.update(entity)

    @JvmDefault
    fun <E : Any> update(entityType: KClass<E>): Update<out Scalar<Int>> = dataStore.update<E>(entityType)

    @JvmDefault
    fun <E : Any> updateAll(entities: Iterable<E>): List<E> = dataStore.update<E>(entities).toList()

    @JvmDefault
    fun <E : Any> delete(): Deletion<out Scalar<Int>> = dataStore.delete()

    @JvmDefault
    fun <E : Any> delete(entityType: KClass<E>): Deletion<out Scalar<Int>> = dataStore.delete<E>(entityType)

    @JvmDefault
    fun <E : Any> delete(entity: E) {
        dataStore.delete(entity)
    }

    @JvmDefault
    fun <E : Any> deleteAll(entities: Iterable<E>) {
        dataStore.delete<E>(entities)
    }

    @JvmDefault
    fun <E : Any> deleteAll(entityType: KClass<E>): Int = dataStore.delete<E>(entityType).get().value()

    @JvmDefault
    fun <E : Any> count(entityType: KClass<E>): Selection<out Scalar<Int>> =
        dataStore.count(entityType)

    @JvmDefault
    fun <E : Any> count(vararg attributes: QueryableAttribute<Any, *>): Selection<out Scalar<Int>> =
        dataStore.count(*attributes)


    @Suppress("UNCHECKED_CAST")
    @JvmDefault
    fun <E : Any> count(entityType: KClass<E>, whereClause: QueryElement<out Result<E>>): Int {

        val query = select(Count.count(entityType.java))
            .applyWhereConditions(whereClause.whereElements)
            .unwrapQuery() as QueryElement<out Result<Tuple>>

        val tuple = query.get().first()
        return tuple.get(0)
    }

    @JvmDefault
    fun <E : Any> exists(entityType: KClass<E>, whereClause: QueryElement<out Result<E>>): Boolean =
        whereClause.limit(1).get().firstOrNull() != null

    @JvmDefault
    fun raw(query: String, vararg parameters: Any): Result<Tuple> =
        dataStore.raw(query, *parameters)

    @JvmDefault
    fun <E : Any> raw(entityType: KClass<E>, query: String, vararg parameters: Any): Result<E> =
        dataStore.raw(entityType, query, *parameters)

    @JvmDefault
    fun <V> runInTransaction(callable: () -> V): V = runInTransaction(null, callable)

    @JvmDefault
    fun <V> runInTransaction(isolation: TransactionIsolation?, callable: () -> V): V =
        when(isolation) {
            null -> dataStore.withTransaction { callable.invoke() }
            else -> dataStore.withTransaction(isolation) { callable.invoke() }
        }

    @JvmDefault
    fun <V> withTransaction(block: (KotlinEntityDataStore<Any>) -> V): V = withTransaction(null, block)

    @JvmDefault
    fun <V> withTransaction(isolation: TransactionIsolation?, block: (KotlinEntityDataStore<Any>) -> V): V =
        when(isolation) {
            null -> dataStore.withTransaction { block.invoke(dataStore) }
            else -> dataStore.withTransaction(isolation) { block.invoke(dataStore) }
        }
}